Type System API refactor

christianparpart opened the pull request:
### The Problem
We could create Solidity source code that caused the solidity compiler to leak memory due to circular dependencies in the type system and no clear ownership.

### The Motivation
With the aim for zero memory leaks I propose to create an API that
* provides clear ownership
* unified way of requesting for types.
* potentially internal optimizations by reusing types or having no heap allocation at all.
* fixes #4517

Implicitly we found out that we can get rid of `shared_ptr<>` and its `shared_from_this()`, overall making the related code look more clear.

### Implementation
The API for requesting/providing types is called `TypeProvider` and lives in `libsolidity/ast/`. It was attempted to keep the changes to all the type accessing code at a minimum, i.e. there were places that I would have written differently (such as how to you access the type provider?), but those should be done in a follow-up PR.

### Notes
* Currently the type provider is accessed via singleton pattern, a follow-up PR should change that. (not included here to keep changes to a minimum!).
* `TypePointer`, `TypePointers`, `FunctionTypePointer` are defined in many places. should be unified (isolated PR)
* I personally don't think the type system should know about the AST / syntax, but currently it does need that information. I'd favor abstracting that part away to keep isolation clean (and avoid circular API layer dependencies) - opinions, @chriseth?
* The type system also implements evaluation via `unaryOperatorResult()` and `binaryOperatorResult()`. I am not quite sure it should be mixed up (if changing at all, then in another PR of course). Objections?
* running soltest/isoltest with `CXXFLAGS=-fsanitize=address -fno-omit-frame-pointer` (yields no errors but without it does(?)).

p.s.: this branch is yet to be squashed when considered ready.

stackenbotten commented:
There was an error when running `test_check_style` for commit `2a926ccda8e8f971e3856a614d85d539adaab6bf`:
```
Error: Trailing whitespace found:
 libdevcore/ArrayUtil.h:54: 

```
Please check that your changes are working as intended.

stackenbotten commented:
There was an error when running `test_check_style` for commit `fb10a16e669fa5b38e72a6454cbc034d561592bc`:
```
Coding style error:
 libdevcore/ArrayUtil.h:46: static_assert(conjunction<not_ref_wrapper<Types>...>::value,

```
Please check that your changes are working as intended.

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,235 @@
+	AddressType const* payableAddressType() const noexcept { return &m_payableAddressType; }
+	AddressType const* addressType() const noexcept { return &m_addressType; }
+
+	IntegerType const* intType(unsigned m = 256) const
```
```suggestion
	IntegerType const* intType(unsigned _bits = 256) const
```

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,235 @@
+
+	StringLiteralType const* stringLiteralType(std::string const& literal);
+
+	TupleType const* tupleType(std::vector<Type const*>&& members);
```
Why not just pass by value?

christianparpart replied:
I agree with having this question. I first checked what cases we have on the caller side, and none of them re-used the passed parameter after the tupleType() call. Passing by value though would give us the freedom to allow both cases, I agree.

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,355 @@
+
+ArrayType const* TypeProvider::arrayType(DataLocation _location, bool _isString)
+{
+	// TODO: reuse existing types
```
I think this will be really hard. Until we do that, it might be better to just have a single vector of `Type*` instead of different vectors of different types. Even then we can even implement `operator<` and first discriminate on category first.

chriseth replied:
Also, create, add to set, return could be implemented as a template for all (easy to construct) types at a time.

chriseth commented:
That's great news! I wonder whether removing the singleton is really necessary. In the end, this is a memory management class...

ekpyron commented:
Just out of curiosity: If this is done properly and reliably everywhere, we could implement ``Type::operator==`` simply as
``bool Type::operator==(Type const& _other) const { return this == &_other; }``, right? Or would that be invalid/undefined due to the types being polymorphic anyways? I'm not entirely sure :-). There shouldn't be two instances of types comparing as equal, though, should there?

christianparpart commented:
@ekpyron that was also my initial thought of having that implicit benefit. But currently we still allow type instanciation outsite `TypeProvider`. With that being prohibited **and** having made sure that `TypeProvider` returns the same type for the same input data, then it would be a simple pointer comparison (or am I missing something?).
p.s.: such a guarantee could be implemented in a later PR.

chriseth commented:
I think it is safer to compare by contents - I don't expect too much gain from comparing by pointer.

ekpyron commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,213 @@
+	/**
+	 * Global TypeProvider instance.
+	 */
+	static TypeProvider& get()
```
I wonder whether it'd be nicer to make this one private and all the ones below static as well and let them use this one - e.g.
``static BoolType const* boolType() const noexcept { return &get().m_boolType; }``
and then
``auto boolType = TypeProvider::boolType();`` instead of ``auto boolType = TypeProvider::get().boolType();``

chriseth replied:
I agree - if we stay with singleton, then it would be better to avoid the `get` call.

ekpyron commented on libsolidity/ast/ASTAnnotations.h:
```diff
@@ -45,7 +45,7 @@ namespace solidity
 
 class Type;
-using TypePointer = std::shared_ptr<Type const>;
+using TypePointer = Type const*;
```
Actually, do you happen to know, if - given this alias - ``TypePointer x;`` is guaranteed to be zero-initialized :-)? (That is in contrast to plain ``Type const* x;`` which will *not* be zero-initialized) I vaguely remember that there might be a dark corner in the standard about this, but I'm not sure anymore and I can't find it right now...
When I try this experimentally, this seems to be actually true, but I wouldn't trust that without a standard reference, try:
```
    using DefaultZeroIntPtr = int const*;
    int const* a;
    DefaultZeroIntPtr b;
    std::cout << a << std::endl << b << std::endl;
```

I'm mentioning this, because if this is true, then you could skip all member initializations of ``TypePointer`` :-).

christianparpart replied:
after short local discussion, we concluded to leave TypePointer var-decls explicitly initialized.

chriseth commented on libdevcore/ArrayUtil.h:
```diff
@@ -0,0 +1,61 @@
+
+} // namespace details
+
+/// C++14 compatible implementation of std::make_array() from C++17.
```
The file could be named `make_array.h`

chriseth replied:
And the function could also use the stl naming scheme

christianparpart replied:
I explicitly chose not to name it `make_array` to not have instantly compilation errors when one choses to compile in C++17 mode and the compiler couldn't chose which `make_array` to use.

Hmm.. I'm changing it.

chriseth commented on test/libsolidity/AnalysisFramework.h:
```diff
@@ -35,8 +35,8 @@ namespace solidity
 class FunctionType;
-using TypePointer = std::shared_ptr<Type const>;
-using FunctionTypePointer = std::shared_ptr<FunctionType const>;
+using TypePointer = Type const*;
```
This is a critical change. It changes a non-POD to a POD type, which means that local variables of TypePointer type are no longer guaranteed to be zero-initialized. Did you check if that can lead to problems?

christianparpart replied:
The compiler actually told me all those problems, tbh. After I found out that the compiler told me some spots I started searching with REs to all `TypePointer` variable declarations and made sure they're `nullptr` initialized if not explicitly initialized right after.

chriseth commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -64,6 +65,11 @@ using namespace dev;
 
+CompilerStack::~CompilerStack()
+{
+	TypeProvider::get().reset();
```
This basically means that `TypeProvider` is "owned" by CompilerStack. Can we make that more explicit? For example in that the CompilerStack "creates" the singleton and thus ensures that a) it is not used before CompilerStack is instanciated and b) it is not "owned" by more than one CompilerStack?

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,395 @@
+	return makeArray<FixedBytesType>(FixedBytesType(static_cast<unsigned>(N) + 1)...);
+}
+
+BoolType const TypeProvider::m_boolType{};
```
Is this required to be mentioned here?

christianparpart replied:
static class variables must be instanciated somewhere, as they cannot be initialized in the class declaration.

christianparpart replied:
actually, just checked. if BoolType() (etc) would be `constexpr`, then I could have put it into the header and skip the explicit instanciation outside the class decl here.

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,395 @@
+
+void TypeProvider::reset()
+{
+	clearCache(
```
Whenever we add a new type, we have to also add it here. Wouldn't it be better to just re-create the TypeProvider?

christianparpart replied:
Recreating the type provider actually would implicitly clear that concern, yes. And that was my initial idea. However, I'd favor doing that in an isolated PR as a follow-up stage.

ekpyron commented:
Actually, I'm still not entirely convinced by this concept.
Which parts of the types have to be reset at all? Is there anything else than the member lists? They depend on the ``ContractDefinition`` and IMO should thus be moved there (in a map from types to member lists) - only the ``ContractDefinition const* = nullptr`` case should be in the type and that remains constant anyways.

If there's no other case, then this means that now all types can be singleton themselves, which means one can make constructors private and have static creation function in each of them, which will immediately prevent "stray types" (like we had to track them down for hours in this PR ;-)).

And if one worries about memory usage after compiling multiple contracts one can still add ``clearTypeCache`` functions to the potentially "large" types, i.e. the reference types. (EDIT: and maybe tuple types; such clearing would need to be done very carefully, since no references to cleared types should remain, but that's not more dangerous than the current clearing and it'd not be necessary in the first step (or maybe even not ever) anyways).

I think this would both be cleaner and safer and in fact mean less invasive code changes.

So @christianparpart, unless you have a compelling argument against that and for the approach of this PR I'd rather suggest to do it that way, sorry ;-).

ekpyron commented:
So the alternative proposal would be:

- move the member lists from the types to ``ContractDefinition`` (resp. it's annotation). Then all types should remain globally valid for the entire lifetime of the compiler (to be verified!).
- make the constructors of all types private
- have a static list/map of already created types of that kind in each type with constructors (either as static members of the type or locally static in the creation function, see next point)
- have static creation functions in each type that will check whether a matching type was already created and return it, or create a new one and store it

This way management of particular types will remain local to the respective type (which IMO increases maintainability) and each type is created only once and never destroyed.

The only argument against that would be ever increasing memory usage in long-living compilers. This can still be addressed in a separate step, though, either by clearing the type lists in each type separately or by moving the lookup and storage logic for already created types *only* to a type provider class (as opposed to the whole creation and management logic).

chriseth commented:
I'm not a big fan of moving the member lists out of the type system into the AST, since the members are not a feature of the AST. Also I'm not sure if checking whether a type has been created already or not is a good idea - then the equality checks have to be really strict and correct. Furthermore, what is the big benefit?

ekpyron commented:
The member list *is* a feature of the AST IMO (at least the library imported part of it)... The "native members" can still live in the types themselves, but since the library imported members in fact depend on the AST node and not on the type, I'd manage that there (so have a type->nonNativeMember map in ContractDefinition instead of a ContractDefinition->nonNativeMember map in the type).
Actually, since in my proposal all types live forever anyways, we could in fact keep it the way it is - I think it's only a problem due to the clearing in the Type Provider. Still I think it should be moved.

Checking whether a type was already created obviously saves memory in the long run, but it's optional in both this PR and my proposal. Shouldn't the equality checks be strict and correct anyways? But yeah - in both proposals we can first just always create a new type and think about saving memory later.

chriseth commented:
Can you elaborate on the native / non-native distinction?

ekpyron commented:
For example ``[address].send`` is native to the type. It's always there independently of where the address occurs. Given ``using L for address;`` all these library functions are non-native to the type - they are only there, if the address is used in scope of the "using".
``[address].send`` could remain in the type, but the additional members stemming from library imports should be managed in the contract definitions, i.e. the scopes of the using definitions.

chriseth commented:
Ok, I see. So this is only about the `using x for y` statements, right? The non-native members would depend on the context and the type. I think it would be OK to store that in the annotations, but the problem is that this is a potentially unbounded container which would rather pollute the annotations. 

Perhaps it would make sense to create a class specifically for that purpose that just serves as a cache and is reset at regular intervals?

ekpyron commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,224 @@
+		static TypeProvider _provider;
+		return _provider;
+	}
+
```
To argue my point: the next 150 lines or so are in fact duplicates of the ones in ``Types.h`` and that's only the declarations, not the definitions - adding or changing any constructor will now require at least four edits in two different files... moving all that to the types will probably not require less edits in the end, but they are at least in the same place. (One could use some template construction to mitigate all that either way, but still I think this stuff actually belongs in the types)

ekpyron commented:
@chriseth Is it better to pollute the types instead of the AST node annotations (which is what's currently happening) :-)? But yeah - we could have a dedicated cache for the member lists that can be cleared.

ekpyron commented:
In any case: the main question is whether we want to stay with the TypeProvider construction as currently done in this PR or whether we want to let the Types manage themselves as I suggested. I'm not strongly against the current TypeProvider construction, but I'd prefer it the other way.

chriseth commented:
Yeah, I agree - factory functions inside the types themselves certainly allow for more local changes.

What do we do about RationalNumberType?

ekpyron commented:
``RationalNumberType``s are a bit of a problem either way (unless we're fine with keeping them alive, although they're only used once or so)... Not sure how best to deal with them.

chriseth commented:
This sounds like we need a singleton cache that can be reset from the compiler stack :)

ekpyron commented:
If we only reset in the compiler stack, then we can just reset the singleton in ``RationalNumberType`` itself :-) - I was wondering if we need to keep the ``RationalNumberType``s alive even that long (I could imagine that a lot of them could be very short-lived, but I'd need to check).

A singleton container that can be reset could be very simple, if we don't want to check, whether an identical type was already created, but always create new ones anyways, e.g.:

```
class TypeContainer {
public:
    template<typename T>
    T *storeType(std::unique_ptr<T>&& _type) {
        T* ptr = _type.get();
        m_types.emplace_back(std::move(_type));
        return ptr;
    }
    void reset() { m_types.clear(); }
    static TypeContainer& get() {
        static TypeContainer container;
        return container;
    }
private:
    std::vector<std::unique_ptr<Type>> m_types;
}
// In Types.h/cpp
class SomeType {
public:
    template<typename... Args>
    static SomeType *someType(Args... args) {
        return TypeContainer::get().storeType(std::make_unique<SomeType>(std::forward<Args>(args)...));
    }
private:
    SomeType(SomeArgs... args) { ... }
};
```
And that could still be extended with equality checks later on, while keeping the creation logic in the types...

ekpyron commented:
But yeah - whether we have all creation functions grouped together in a separate class or keep them in each Type itself is mostly a question of taste in the end - so I'm fine with keeping it as it is as well, especially since that's already implemented here.

chriseth commented:
The critical thing is that ownership should be clear. Types are stored as pointers and the lifetime of these pointers has to be bounded by calls to reset - at least for everything that is not just a cache. The members might be just a cache if they do not store types themselves. Every type that is not "static", like RationalNumberType has to have clear ownership and lifetime, and I think that can only be provided by the CompilerStack.

ekpyron commented:
Actually, to me it looks like two cases then: If it's feasible - instead of a singleton that is reset from ``CompilerStack`` - to make ``TypeProvider`` an actual member of ``CompilerStack``, I'd say that's preferrable and would vote for basically keeping things as they are (we can try mitigating code duplication, resp. spread later with some template tricks).
If having ``TypeProvider`` be a member results in too much hassle in passing it around everywhere and we would end up keeping it a singleton, then I would change it to just be a thin (singleton) ``TypeContainer`` as in my comment above and have the creation functions statically in the types themselves for increased locality.

Commit:
[test] AnalysisFramework's instance of CompilerStack to be lazily instanciated.

Commit:
libsolidity: Introducing TypeProvider API, for clear type system ownership.

chriseth commented on libsolidity/analysis/GlobalContext.cpp:
```diff
@@ -34,42 +35,50 @@ namespace dev
-		StateMutability::Pure
-	)),
-})
+std::vector<std::shared_ptr<MagicVariableDeclaration const>> constructMagicVariables()
```
Please move into anonymous namespace.

chriseth replied:
Also no need for `std::`.

chriseth commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -195,7 +196,7 @@ TypePointers TypeChecker::typeCheckABIDecodeAndRetrieveReturnType(FunctionCall c
 		{
 			m_errorReporter.typeError(typeArgument->location(), "Argument has to be a type name.");
-			components.push_back(make_shared<TupleType>());
+			components.push_back(TypeProvider::errorType());
```
Why not just `emptyTupleType`? In which way is it an error type?

chriseth commented on libsolidity/analysis/TypeChecker.h:
```diff
@@ -28,6 +28,7 @@
 #include <libsolidity/ast/ASTForward.h>
 #include <libsolidity/ast/ASTVisitor.h>
 #include <libsolidity/ast/Types.h>
+#include <libsolidity/ast/TypeProvider.h>
```
Do we really need this include in the header?

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,229 @@
+
+#pragma once
+
+#include <libsolidity/ast/AST.h>
```
ASTForward could be enough.

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,229 @@
+	TypeProvider& operator=(TypeProvider const&) = delete;
+	~TypeProvider() = default;
+
+	/**
```
We usually only use `/**` for classes and `///` for functions.

chriseth commented:
(not yet fully reviewed)

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,228 @@
+	/// suffix " storage", " calldata" or " memory" to a type pointer. If suffix not given, defaults to " storage".
+	static TypePointer fromElementaryTypeName(std::string const& _name);
+
+	/// @returns boolean type.
```
```suggestion
```

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,228 @@
+	static FixedBytesType const* byteType() { return fixedBytesType(1); }
+	static FixedBytesType const* fixedBytesType(unsigned m) { return &m_bytesM.at(m - 1); }
+
+	static ArrayType const* bytesType() noexcept { return &m_bytesType; }
```
I think we should force explicit data location just as we do in Solidity now.

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,228 @@
+	static ArrayType const* stringMemoryType() noexcept { return &m_stringMemoryType; }
+
+	/// Constructor for a byte array ("bytes") and string.
+	static ArrayType const* arrayType(DataLocation _location, bool _isString = false);
```
This should be removed at a later point, I think. It is way too confusing.

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,228 @@
+	static TupleType const* emptyTupleType() noexcept { return &m_emptyTupleType; }
+
+	/// @returns a suitably simple type when a type is expected but an error has occurred.
+	static TupleType const* errorType() noexcept { return emptyTupleType(); }
```
I'm not sure we should extend the semantics of the TypeChecker to the type system - the TypeChecker should just use `emptyTupleType`.

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,228 @@
+		return _type;
+	}
+
+	/// @returns the type of a function.
```
```suggestion
	/// @returns the internally-facing or externally-facing type of a function.
```

chriseth replied:
Does not change the data location of parameters.

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,228 @@
+	static RationalNumberType const* rationalNumberType(Literal const& _literal);
+
+	static RationalNumberType const* rationalNumberType(
+		rational const& _value,
```
Where does the `rational` symbol come from?

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,228 @@
+
+	static TypeType const* typeType(Type const* _actualType);
+
+	static StructType const* structType(StructDefinition const& _struct, DataLocation _location = DataLocation::Storage);
```
We should probably also remove the default value for the location. Also, why is location the first parameter for arrays but the second for structs?

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,228 @@
+	static MappingType const* mappingType(Type const* _keyType, Type const* _valueType);
+
+private:
+	/**
```
Please use `///`

chriseth commented on libsolidity/ast/TypeProvider.h:
```diff
@@ -0,0 +1,228 @@
+	/**
+	 * Global TypeProvider instance.
+	 */
+	static TypeProvider& get()
```
I think `inst()` or even `instance()` would be better - we already have so many functions called `get`.

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,430 @@
+
+BoolType const TypeProvider::m_boolType{};
+InaccessibleDynamicType const TypeProvider::m_inaccessibleDynamicType{};
+ArrayType const TypeProvider::m_bytesType{DataLocation::Storage, false};
```
```suggestion
ArrayType const TypeProvider::m_bytesStorageType{DataLocation::Storage, false};
```

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,430 @@
+InaccessibleDynamicType const TypeProvider::m_inaccessibleDynamicType{};
+ArrayType const TypeProvider::m_bytesType{DataLocation::Storage, false};
+ArrayType const TypeProvider::m_bytesMemoryType{DataLocation::Memory, false};
+ArrayType const TypeProvider::m_stringType{DataLocation::Storage, true};
```
```suggestion
ArrayType const TypeProvider::m_stringStorageType{DataLocation::Storage, true};
```

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,430 @@
+	// MetaType is stored separately
+};
+
+TypeProvider::TypeProvider()
```
Move to header?

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,430 @@
+
+Type const* TypeProvider::fromElementaryTypeName(ElementaryTypeNameToken const& _type)
+{
+	solAssert(TokenTraits::isElementaryTypeName(_type.token()),
```
```suggestion
	solAssert(
	    TokenTraits::isElementaryTypeName(_type.token()),
```

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,430 @@
+
+	return map.emplace(
+		make_pair(m, n),
+		make_unique<FixedPointType>(
```
```suggestion
		make_unique<FixedPointType>(m, n, _modifier)
```

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,430 @@
+		if (_location == DataLocation::Storage)
+			return stringType();
+		if (_location == DataLocation::Memory)
+			return stringMemoryType();
```
What about calldata?

christianparpart replied:
it's just an optimization. the catch-all is below. (as discussed just right now)

chriseth commented on libsolidity/ast/TypeProvider.cpp:
```diff
@@ -0,0 +1,430 @@
+		if (_location == DataLocation::Storage)
+			return bytesType();
+		if (_location == DataLocation::Memory)
+			return bytesMemoryType();
```
What about calldata?

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -45,10 +45,11 @@ namespace dev
 namespace solidity
```
Can delete boost::noncopyable include above.

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -148,10 +150,16 @@ static_assert(std::is_nothrow_move_constructible<MemberList>::value, "MemberList
+	Type(Type const&) = delete;
+	Type(Type&&) = default;
+	Type& operator=(Type const&) = delete;
+	Type& operator=(Type&&) = default;
```
Types should only be used as pointers, so maybe disallow move?

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -382,17 +378,22 @@ class IntegerType: public Type
 	};
 
-	static IntegerType& uint256() { static std::shared_ptr<IntegerType> uint256(std::make_shared<IntegerType>(256)); return *uint256; }
+public:
```
We already had `public` above.

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -382,17 +378,22 @@ class IntegerType: public Type
 	Category category() const override { return Category::Integer; }
 
-	explicit IntegerType(unsigned _bits, Modifier _modifier = Modifier::Unsigned);
+	IntegerType(IntegerType&&) = default;
```
Do we really need to mention those?

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -382,17 +378,22 @@ class IntegerType: public Type
 
-	explicit IntegerType(unsigned _bits, Modifier _modifier = Modifier::Unsigned);
+	IntegerType(IntegerType&&) = default;
+	IntegerType& operator=(IntegerType&&) = default;
```
`m_bits` is `const` so this cannot be moved anyway.

Commit:
TypeProvider: storageType() to make all parameters explicit

chriseth commented on libsolidity/ast/Types.h:
```diff
@@ -942,12 +948,16 @@ class EnumType: public Type
+
 	Category category() const override { return Category::Tuple; }
-	explicit TupleType(std::vector<TypePointer> const& _types = std::vector<TypePointer>()): m_components(_types) {}
+	TupleType(TupleType&&) = default;
```
Can be removed.

chriseth commented on Changelog.md:
```diff
@@ -14,6 +14,7 @@ Compiler Features:
  * SMTChecker: Support arithmetic compound assignment operators.
  * SMTChecker: Support unary increment and decrement for array and mapping access.
  * SMTChecker: Show unsupported warning for inline assembly blocks.
+ * SMTChecker: Support mod.
```
Shouldn't be here.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -1503,18 +1388,32 @@ u256 BoolType::literalValue(Literal const* _literal) const
 	if (_operator == Token::Delete)
-		return TypeResult(make_shared<TupleType>());
-	return (_operator == Token::Not) ? shared_from_this() : TypePointer();
+		return TypeProvider::emptyTupleType(); // TODO: Lag of understanding.
```
```suggestion
		return TypeProvider::emptyTupleType();
```

chriseth replied:
`delete` returns the empty tuple because the type of the operator result is "nothing", i.e. you cannot use it inside any other expression.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -1550,36 +1449,44 @@ bool ContractType::isPayable() const
-	return _operator == Token::Delete ? make_shared<TupleType>() : TypePointer();
+		return nullptr;
+	else if (_operator == Token::Delete)
+		return TypeProvider::emptyTupleType(); // TODO: Lag of understanding.
```
```suggestion
		return TypeProvider::emptyTupleType();
```

Commit:
libsolidity: Be more restrictive to not allow any copy constructors at all in Type system.

This enforces us to be a little bit more historical in how to initialize
our std::array<>'s in TypeProvider.

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -988,10 +989,9 @@ void CompilerUtils::convertType(
-					targetType = dynamic_pointer_cast<StructType const>(targetType.shared_from_this())
-				](CompilerContext& _context) {
+				auto conversionImpl =
+					[typeOnStack = &typeOnStack, targetType = &targetType](CompilerContext& _context)
```
The previous code used `_typeOnStack`

christianparpart replied:
@chriseth: Actually, this code is right.

* the outer-scope `typeOnStack` is a reference type (which cannot be captured in lambdas) whereas the inner `typeOnStack` is a pointer type.
* It's not using the `_typeOnStack` as the original code did before along with a dynamic_cast to StructType, as the outer scope's `typeOnStack` already has been cast to that type.

chriseth replied:
It is confusing anyway...

Commit:
[libsolidity] TypeProvider: adds explicit uint256() accessor and removes default params in integerType(...).

chriseth committed:
Change pre-created types to unique_ptr to avoid copying and moving.

chriseth commented:
Segmentation fault in syntax tests... :(

chriseth commented:
Segfault results from dependencies between static init, I think.

chriseth committed:
Lazy-initialize string and bytes types.

chriseth committed:
Make data location explicit.

chriseth commented:
Hopefully fixed now.

axic commented:
Was this running the asan test and there were no leaks found?

chriseth commented:
@axic yes, please see https://github.com/ethereum/solidity/pull/6537

chriseth commented:
Weird, it worked on develop: https://circleci.com/gh/ethereum/solidity/111617

christianparpart commented:
i'm checking again (it used to be fixed when I checked. but a lot has been done in the middle of this PR, sorry) :-)
